From 70d7fc3df1bb9e1d04fd5688f8801582f50e715e Mon Sep 17 00:00:00 2001
From: Manish Dharanenthiran <quic_mdharane@quicinc.com>
Date: Mon, 27 May 2024 15:10:11 +0530
Subject: [PATCH 5/6] wifi: cfg80211: Fix resetting ssid len and nlportid in
 wdev

With dynamic link removal in place, resetting conn_owner_nlportid and
ssid_len in wireless dev structure is not correct as there might be some
partner links still up and running.

Hence, add changes to prevent resetting the above mentioned parameter
only when all of the links is removed. For legacy cases, the valid_links
param will be always 0, hence the changes will not affect those cases.

Signed-off-by: Manish Dharanenthiran <quic_mdharane@quicinc.com>
---
 net/wireless/ap.c   | 11 +++++++++--
 net/wireless/util.c |  4 ++--
 2 files changed, 11 insertions(+), 4 deletions(-)

--- a/net/wireless/ap.c
+++ b/net/wireless/ap.c
@@ -29,19 +29,31 @@ static int ___cfg80211_stop_ap(struct cf
 	    dev->ieee80211_ptr->iftype != NL80211_IFTYPE_P2P_GO)
 		return -EOPNOTSUPP;
 
-	if (!wdev->links[link_id].ap.beacon_interval)
+	if (!wdev->links[link_id].ap.beacon_interval){
+		if (hweight16(wdev->valid_links) <= 1) {
+			wdev->conn_owner_nlportid = 0;
+			wdev->u.ap.ssid_len = 0;
+		}
 		return -ENOENT;
+	}
 
 	cfg80211_update_last_available(wdev->wiphy,
 				       &wdev->links[link_id].ap.chandef);
 
 	err = rdev_stop_ap(rdev, dev, link_id);
 	if (!err) {
-		wdev->conn_owner_nlportid = 0;
 		wdev->links[link_id].ap.beacon_interval = 0;
 		memset(&wdev->links[link_id].ap.chandef, 0,
 		       sizeof(wdev->links[link_id].ap.chandef));
-		wdev->u.ap.ssid_len = 0;
+
+		if (hweight16(wdev->valid_links) <= 1) {
+			/* Clear this only when there is one or lesser valid
+			 * link, otherwise consider that some link is present
+			 */
+			wdev->conn_owner_nlportid = 0;
+			wdev->u.ap.ssid_len = 0;
+		}
+
 		rdev_set_qos_map(rdev, dev, NULL);
 		if (notify)
 			nl80211_send_ap_stopped(wdev, link_id);
